docbrown/db/src/graphgen/preferential_attachment.rs

```diff
@@ -0,0 +1,93 @@
+use crate::graph::Graph;
+use rand::prelude::*;
+use docbrown_core::Direction;
+use std::collections::HashSet;
+
+//let mut weights:Vec<usize> =view.vertices().map(|v|v.degree()).collect();
+
+pub fn preferential_attachment(graph:&Graph,vertices_to_add:usize,edges_per_step:usize){
+
+    let mut rng = rand::thread_rng();
+    let mut latest_time = graph.latest_time();
+
+    let view = graph.window(i64::MIN,i64::MAX);
+    let mut ids:Vec<u64> = view.vertex_ids().collect();
+    let mut degrees:Vec<usize> = view.vertices().map(|v| v.degree()).collect();
+    let mut edge_count:usize = degrees.iter().sum();
+
+    let mut max_id = match ids.iter().max() {
+        Some(id) => {*id},
+        None=>0
+    };
+
+    if(ids.len()==0){
+        graph.add_vertex(latest_time,0,&vec![]);
+        degrees.push(0);
+        ids.push(0);
+    }
+    while ids.len() < edges_per_step {
+        max_id+=1;
+        latest_time+=1;
+        let prev_node = ids[ids.len()-1];
+        graph.add_vertex(latest_time,max_id,&vec![]);
+        graph.add_edge(latest_time,max_id,prev_node,&vec![]);
+        degrees[ids.len()-1]+=1;
+        degrees.push(1);
+        ids.push(max_id);
+        edge_count+=2;
+
+    }
+
+
+    for i in 0..vertices_to_add {
+        max_id += 1;
+        latest_time += 1;
+        let mut normalisation = edge_count.clone();
+        let mut positions_to_skip: HashSet<usize> = HashSet::new();
+
+        for j in 0..edges_per_step {
+            let mut sum = 0;
+            let rand_num = rng.gen_range(1..=normalisation);
+            for(pos,id) in ids.iter().enumerate() {
+                if ! positions_to_skip.contains(&pos){
+                    sum += degrees[pos];
+                    if sum>= rand_num {
+                        positions_to_skip.insert(pos);
+                        normalisation-degrees[pos];
+                        break;
+                    }
```

Comment:
Not sure this is the most efficient sampler though it does look ok. I suppose the structure of the preferential attachment helps a lot here to keep this somewhat efficient as nodes with large degree should be near the beginning of the list...

Reply:
I am sure there is a better way of doing this out there, however, there doesn't seem to be any decent data structures for maintaining a weighted index for sampling. For those that do exist, you can only update the weights not add new elements to the structure without rebuilding. These often don't allow for sampling without replacement as well which is key here.